openupgrade_scripts/scripts/hr_holidays/18.0.1.6/pre-migration.py

```diff
@@ -0,0 +1,194 @@
+            JOIN hr_employee_{table}_rel rel ON rel.{table}_id = leave.id
+            WHERE leave.holiday_type = 'employee' AND leave.multi_employee
+            GROUP BY leave.id
+            """
+        )
+        leave_employees.extend(env.cr.fetchall())
+        # case company
+        env.cr.execute(
+            f"""
+            SELECT leave.id, 'company', array_agg(he.id) as employee_ids
+            FROM {table} AS leave
+            JOIN hr_employee he ON he.company_id = leave.mode_company_id
+            WHERE leave.holiday_type = 'company'
+            GROUP BY leave.id
+            """
+        )
+        leave_employees.extend(env.cr.fetchall())
+        # case category
+        env.cr.execute(
+            f"""
+            SELECT leave.id, 'category', array_agg(rel.employee_id) as employee_ids
+            FROM {table} AS leave
+            JOIN employee_category_rel rel ON rel.category_id = leave.category_id
+            WHERE leave.holiday_type = 'category'
+            GROUP BY leave.id
+            """
+        )
+        leave_employees.extend(env.cr.fetchall())
+        # case department
+        env.cr.execute(
+            f"""
+            SELECT leave.id, 'department', array_agg(he.id) as employee_ids
+            FROM {table} AS leave
+            JOIN hr_employee he ON he.department_id = leave.department_id
+            WHERE leave.holiday_type = 'department'
+            GROUP BY leave.id
+            """
+        )
+        leave_employees.extend(env.cr.fetchall())
+        for table_id, holiday_type, employee_ids in leave_employees:
+            employees = env["hr.employee"].browse(employee_ids)
+            if employee_ids:
+                openupgrade.logged_query(
+                    env.cr,
+                    f"""
+                    UPDATE {table}
+                    SET employee_id = {employees[0].id}
+                    WHERE id = {table_id}
+                    """,
+                )
+                if holiday_type != "employee":
+                    openupgrade.logged_query(
+                        env.cr,
+                        f"""
+                        UPDATE {table}
+                        SET holiday_type = 'employee'
+                        WHERE id = {table_id}
+                        """,
+                    )
+                for employee in employees[1:]:
+                    openupgrade.logged_query(
+                        env.cr,
+                        f"""
+                        INSERT INTO {table} (employee_id, {columns})
+                        SELECT {employee.id}, {columns}
+                        FROM {table}
+                        WHERE id = {table_id}
+                        """,
+                    )
+            else:
+                openupgrade.logged_query(
+                    env.cr,
+                    f"""
+                    DELETE FROM {table} WHERE id = {table_id}
+                    """,
+                )
```

Comment:
Hi @MiquelRForgeFlow are you sure about that one ?
On my database, everytime I had a leave or allocation concerning multiple employees, individuals leave or allocations were also created for each one of these.
Now running migration on my database, I end up with duplicate leave and allocations (even worse when I had allocation for company, it created allocations for employees which are not anymore under contract for years).

Maybe we should do that only for leave / allocation which are in draft state only ?